File: p203e/remove_elements.cpp
// g++ -std=c++11 *.cpp -o test && ./test && rm -f test
#include <vector>
#include <iostream>

using namespace std;
/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode** pp = &head;
        while (*pp) {
            if ((*pp)->val == val) { // remove
                ListNode* del = *pp;
                *pp = (*pp)->next;
                delete del;
            } else {
                pp = &((*pp)->next);
            }
        }
        return head;
    }
};
int main(int argc, char const *argv[])
{
    vector<int> testVals = {6,1,2,3,6,4,5,6};
    ListNode* dummy = new ListNode(0);
    ListNode* p = dummy;
    for (int v: testVals) {
        p->next = new ListNode(v);
        p = p->next;
    }

    Solution sol;
    ListNode* after = sol.removeElements(dummy->next, 6);

    cout << "After removal: ";
    p = after;
    while (p != nullptr) {
        cout << p->val << "->";
        p = p->next;
    }
    cout << endl;
    return 0;
}

File: p203e/remove_elements.py
from local_packages.list import ListNode

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        sentinel = ListNode(val + 1, head)
        p = sentinel
        while p and p.next:
            if p.next.val == val:
                p.next = p.next.next
            else:
                p = p.next
        return sentinel.next


# TESTS
tests = [
    ([], 1, []),
    ([1], 2, [1]),
    ([1, 1], 1, []),
    ([6, 1, 2, 3, 6, 4, 5, 6], 6, [1, 2, 3, 4, 5]),
]
for t in tests:
    sol = Solution()
    actual = sol.removeElements(ListNode.from_array(t[0]), t[1])
    print("Remove elements from", t[0], "->", ListNode.to_array(actual))
